src/Npgsql/PoolManager.cs

```diff
@@ -24,11 +24,12 @@ static class PoolManager
         internal static bool TryGetValue(string key, [NotNullWhenTrue] out ConnectorPool? pool)
         {
             // Note that pools never get removed. _pools is strictly append-only.
+            var nextSlot = _nextSlot;
```

Comment:
This local can be moved down to be after `pools`, so it won't separate the comment about `pools` and the variable declaration.

Reply:
@YohDeadfall see https://github.com/npgsql/npgsql/issues/2485#issuecomment-497425470. If `nextSlot` is captured after `pools`, there's again a race condition where it has already been increased, and we still get the IndexOfOutBoundsException. The capturing order here is very important - isn't lock-free programming the best :)

Reply:
Even without my proposal it's still possible, because instruction order doesn't guarantee the same execution order. There should be volatile read.

Reply:
You're right - I didn't think of that. That's an `Interlocked.Exchange()` on the writing side, but nothing on the reading side - @thetranman can you please make this change?

Reply:
Sounds good. Should we also add a local copy of _nextSlot before iterating in PoolManager.ClearAll()?

Reply:
We can do this, although there's going to be almost zero value in doing so. That method is really not perf-sensitive...